python/cucim/src/cucim/core/operations/color/stain_normalizer.py

```diff
@@ -0,0 +1,821 @@
+        If True, a dummy stain vector, orthogonal to the first two will be
+        added.
+
+    Returns
+    -------
+    stain_coeff : cp.ndarray
+        Stain attenuation coefficient matrix derived from the image, where
+        the first column corresponds to H, the second column is E and the rows
+        are RGB values.
+
+    References
+    ----------
+    .. [1] M. Macenko et al., "A method for normalizing histology slides for
+           quantitative analysis," 2009 IEEE International Symposium on
+           Biomedical Imaging: From Nano to Macro, 2009, pp. 1107-1110,
+           doi: 10.1109/ISBI.2009.5193250.
+           http://wwwx.cs.unc.edu/~mn/sites/default/files/macenko2009.pdf
+    """
+
+    _validate_image(image)
+    channel_axis = _prep_channel_axis(channel_axis, image.ndim)
+    if alpha < 0 or alpha > 100:
+        raise ValueError("alpha must be a percentile in range [0, 100].")
+    if beta < 0:
+        raise ValueError("beta must be nonnegative.")
+
+    # convert to absorbance (optical density) matrix of shape (3, n_pixels)
+    absorbance = _image_to_absorbance_matrix(
+        image,
+        source_intensity=source_intensity,
+        image_type=image_type,
+        channel_axis=channel_axis,
+    )
+
+    # remove transparent pixels
+    absorbance = absorbance[:, cp.all(absorbance > beta, axis=0)]
+    if absorbance.size == 0:
+        raise ValueError(
+            "All pixels of the input image are below the threshold."
+        )
+
+    # compute eigenvectors (do small 3x3 matrix calculations on the host)
+    cov = _covariance(absorbance, rowvar=True)
+    cov = cp.asnumpy(cov).astype(np.float32, copy=False)
+    _, ev = np.linalg.eigh(cov)
+    ev = ev[:, [2, 1]]
+    # flip to ensure positive first coordinate so arctan2 angles are about 0
+    if ev[0, 0] < 0:
+        ev[:, 0] *= -1
+    if ev[0, 1] < 0:
+        ev[:, 1] *= -1
+
+    # project on the plane spanned by the eigenvectors
+    projection = cp.dot(cp.asarray(ev.T), absorbance)
+
+    # find the vectors that span the whole data (min and max angles)
+    phi = cp.arctan2(projection[1], projection[0])
+    min_phi, max_phi = cp.percentile(phi, (alpha, 100 - alpha))
+    # need these scalars on the host
+    min_phi, max_phi = float(min_phi), float(max_phi)
+
+    # project back to absorbance space
+    v_min = np.array([math.cos(min_phi), math.sin(min_phi)], dtype=np.float32)
+    v_max = np.array([math.cos(max_phi), math.sin(max_phi)], dtype=np.float32)
+    v1 = np.dot(ev, v_min)
+    v2 = np.dot(ev, v_max)
+
+    # Make Hematoxylin (H) first and eosin (E) second by comparing the
+    # R channel value
+    if v1[0] < v2[0]:
+        v1, v2 = v2, v1
+    stain_coeff = np.stack((v1, v2), axis=-1)
+
+    # renormalize columns to reduce numerical error
+    stain_coeff /= np.linalg.norm(stain_coeff, axis=0, keepdims=True)
+    if append_third_column:
+        stain_coeff = _complement_stain_matrix(stain_coeff)
```

Comment:
Is there any reason to do this calculations on the host?

Reply:
Out of curiousity, I ran the following
```Python
ev = cp.random.randn(3, 3).astype(np.float32)
min_phi = cp.asarray(-.5)
max_phi = cp.asarray(0.5)

def _on_host(cov, min_phi, max_phi):
    cov = cp.asnumpy(cov).astype(np.float32, copy=False)
    _, ev = np.linalg.eigh(cov)
    ev = ev[:, [2, 1]]
    # flip to ensure positive first coordinate so arctan2 angles are about 0
    if ev[0, 0] < 0:
        ev[:, 0] *= -1
    if ev[0, 1] < 0:
        ev[:, 1] *= -1
    min_phi, max_phi = float(min_phi), float(max_phi)

    # project back to absorbance space
    v_min = np.array([math.cos(min_phi), math.sin(min_phi)], dtype=np.float32)
    v_max = np.array([math.cos(max_phi), math.sin(max_phi)], dtype=np.float32)
    v1 = np.dot(ev, v_min)
    v2 = np.dot(ev, v_max)

    # Make Hematoxylin (H) first and eosin (E) second by comparing the
    # R channel value
    if v1[0] < v2[0]:
        v1, v2 = v2, v1
    stain_coeff = np.stack((v1, v2), axis=-1)

    # renormalize columns to reduce numerical error
    stain_coeff /= np.linalg.norm(stain_coeff, axis=0, keepdims=True)
    return cp.asarray(stain_coeff)
    
def _on_device(cov, min_phi, max_phi):

    # project back to absorbance space
    _, ev = cp.linalg.eigh(cov)
    ev = ev[:, [2, 1]]
    # flip to ensure positive first coordinate so arctan2 angles are about 0
    if ev[0, 0] < 0:
        ev[:, 0] *= -1
    if ev[0, 1] < 0:
        ev[:, 1] *= -1    
    v_min = cp.array([cp.cos(min_phi), cp.sin(min_phi)], dtype=np.float32)
    v_max = cp.array([cp.cos(max_phi), cp.sin(max_phi)], dtype=np.float32)
    v1 = cp.dot(ev, v_min)
    v2 = cp.dot(ev, v_max)

    # Make Hematoxylin (H) first and eosin (E) second by comparing the
    # R channel value
    if v1[0] < v2[0]:
        v1, v2 = v2, v1
    stain_coeff = cp.stack((v1, v2), axis=-1)

    # renormalize columns to reduce numerical error
    stain_coeff /= cp.linalg.norm(stain_coeff, axis=0, keepdims=True)
    return cp.asarray(stain_coeff)

```

which gives
```
In [22]: %timeit _on_host(ev, min_phi, max_phi); d.synchronize()
128 µs ± 1.52 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [23]: %timeit _on_device(ev, min_phi, max_phi); d.synchronize()
677 µs ± 119 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

So, approximately 0.5 ms additional overhead regardless of image size when all of these tiny calculations are done on the GPU.